File: Projetos/ACBrLib/Demos/C#/AbecsPinpad/Import/Dinamico/MT/ACBrAbecsPinpad.cs
using System;
using System.Globalization;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using ACBrLib.Core;
using ACBrLib.AbecsPinpad;
using System.Threading;

//MT
namespace ACBrLib.AbecsPinpad
{
    /// <inheritdoc />
    public sealed partial class ACBrAbecsPinpad : ACBrLibHandle
    {

        public ACBrAbecsPinpad(string eArqConfig = "", string eChaveCrypt = "") : base(IsWindows ? "ACBrAbecsPinpad64.dll" : "libacbrabecspinpad64.so",
                                                                                      IsWindows ? "ACBrAbecsPinpad32.dll" : "libacbrabecspinpad32.so")
        {
            var inicializar = GetMethod<AbecsPinpad_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ref libHandle, ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);

            Config = new AbecsPinpadConfig(this);
        }

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<AbecsPinpad_Nome>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<AbecsPinpad_Versao>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public AbecsPinpadConfig Config { get; }


        public override void ConfigGravar(string eArqConfig = "")
        {
            var gravarIni = GetMethod<AbecsPinpad_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override void ImportarConfig(string eArqConfig)
        {
            var lerIni = GetMethod<AbecsPinpad_ConfigImportar>();
            var ret = ExecuteMethod(() => lerIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override string ExportarConfig()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_ConfigExportar>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public override void ConfigLer(string eArqConfig = "")
        {
            var lerIni = GetMethod<AbecsPinpad_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<AbecsPinpad_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public override void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<AbecsPinpad_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        public void Ativar()
        {
            var method = GetMethod<AbecsPinpad_Ativar>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void Desativar()
        {
            var method = GetMethod<AbecsPinpad_Desativar>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void OPN()
        {
            var method = GetMethod<AbecsPinpad_OPN>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void CLO(string sMensagem = "")
        {
            var method = GetMethod<AbecsPinpad_CLO>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(sMensagem)));

            CheckResult(ret);
        }

        public void CLX(string sMensagemOuNomeImagem = "")
        {
            var method = GetMethod<AbecsPinpad_CLX>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(sMensagemOuNomeImagem)));

            CheckResult(ret);
        }

        public string GIX(string PP_DATA)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_GIX>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(PP_DATA), buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string GIN(int GIN_ACQIDX)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_GIN>();
            var ret = ExecuteMethod(() => method(libHandle, GIN_ACQIDX, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string PinPadCapabilities()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_PinPadCapabilities>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public void DSP(string sMensagem)
        {
            var method = GetMethod<AbecsPinpad_DSP>();
            var ret = ExecuteMethod(() => method(libHandle, sMensagem));

            CheckResult(ret);
        }

        public void DEX(string sMensagem)
        {
            var method = GetMethod<AbecsPinpad_DEX>();
            var ret = ExecuteMethod(() => method(libHandle, sMensagem));

            CheckResult(ret);
        }

        public void GKY()
        {
            var method = GetMethod<AbecsPinpad_GKY>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);
        }

        public void RMC(string sMensagemRMC)
        {
            var method = GetMethod<AbecsPinpad_RMC>();
            var ret = ExecuteMethod(() => method(libHandle, sMensagemRMC));

            CheckResult(ret);
        }

        public string GCD(int aMSGIDX, int aTimeOut)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_GCD>();
            var ret = ExecuteMethod(() => method(libHandle, aMSGIDX, aTimeOut, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string CEX(bool VerifyKey, bool VerifyMagnetic, bool VerifyICCInsertion, bool VerifyICCRemoval, bool VerifyCTLSPresence, int aTimeOut)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_CEX>();
            var ret = ExecuteMethod(() => method(libHandle, VerifyKey, VerifyMagnetic, VerifyICCInsertion, VerifyICCRemoval, VerifyCTLSPresence, aTimeOut, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string MNU(string sMNUOPT, string sDSPMSG, int aTimeOut)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_MNU>();
            var ret = ExecuteMethod(() => method(libHandle, sMNUOPT, sDSPMSG, aTimeOut, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string LoadMedia(string sCaminhoImagem, int aTipoImagem)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_LoadMedia>();
            var ret = ExecuteMethod(() => method(libHandle, sCaminhoImagem, aTipoImagem, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string LMF()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<AbecsPinpad_LMF>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public void DSI(string sNomeArquivo)
        {
            var method = GetMethod<AbecsPinpad_DSI>();
            var ret = ExecuteMethod(() => method(libHandle, sNomeArquivo));

            CheckResult(ret);
        }

        public void DMF(string sNomeArquivo)
        {
            var method = GetMethod<AbecsPinpad_DMF>();
            var ret = ExecuteMethod(() => method(libHandle, sNomeArquivo));

            CheckResult(ret);
        }

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<AbecsPinpad_Finalizar>();
            var codRet = ExecuteMethod(() => finalizar(libHandle));
            CheckResult(codRet);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<AbecsPinpad_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
            return FromUTF8(buffer);
        }
    }
}

File: Projetos/ACBrLib/Demos/Java/AbecsPinpad/Imports/ACBrLibAbecsPinpadMT/src/com/acbr/abecspinpad/ACBrAbecsPinpad.java
package com.acbr.abecspinpad;

import com.acbr.ACBrLibBase;
import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import java.io.File;

import java.nio.ByteBuffer;
import java.nio.file.Paths;

public final class ACBrAbecsPinpad extends ACBrLibBase {

    public interface ACBrAbecsPinpadLib extends Library {

        public static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public static ACBrAbecsPinpadLib INSTANCE = LibraryLoader.getInstance();

        class LibraryLoader {

            private static String library = "";
            private static ACBrAbecsPinpadLib instance = null;

            public static String getLibraryName() {
                if (library.isEmpty()) {
                    if (Platform.isWindows()) {
                        library = Platform.is64Bit() ? "ACBrAbecsPinpad64" : "ACBrAbecsPinpad32";
                    } else {
                        library = Platform.is64Bit() ? "acbrabecspinpad64" : "acbrabecspinpad32";
                    }
                }
                return library;
            }

            public static ACBrAbecsPinpadLib getInstance() {
                if (instance == null) {
                    instance = (ACBrAbecsPinpadLib) Native.synchronizedLibrary((Library) Native.load(JNA_LIBRARY_NAME, ACBrAbecsPinpadLib.class));
                }
                return instance;
            }
        }

        int AbecsPinpad_Inicializar(PointerByReference libHandler, String eArqConfig, String eChaveCrypt);

        int AbecsPinpad_Finalizar(Pointer libHandler);

        int AbecsPinpad_Nome(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int AbecsPinpad_Versao(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int AbecsPinpad_UltimoRetorno(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int AbecsPinpad_ConfigImportar(Pointer libHandler, String eArqConfig);

        int AbecsPinpad_ConfigExportar(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int AbecsPinpad_ConfigLer(Pointer libHandler, String eArqConfig);

        int AbecsPinpad_ConfigGravar(Pointer libHandler, String eArqConfig);

        int AbecsPinpad_ConfigLerValor(Pointer libHandler, String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int AbecsPinpad_ConfigGravarValor(Pointer libHandler, String eSessao, String eChave, String valor);

        int AbecsPinpad_Ativar(Pointer libHandler);

        int AbecsPinpad_Desativar(Pointer libHandler);
        
        int AbecsPinpad_OPN(Pointer libHandler);
        
        int AbecsPinpad_CLO(Pointer libHandler, String sMensagem);
        
        int AbecsPinpad_CLX(Pointer libHandler, String sMensagemOuNomeImagem);
        
        int AbecsPinpad_GIX(Pointer libHandler, String PP_DATA, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_GIN(Pointer libHandler, int GIN_ACQIDX, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_PinPadCapabilities(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_DSP(Pointer libHandler, String sMensagem);
          
        int AbecsPinpad_DEX(Pointer libHandler, String sMensagem);
          
        int AbecsPinpad_GKY(Pointer libHandler);
          
        int AbecsPinpad_RMC(Pointer libHandler, String sMensagemRMC);
          
        int AbecsPinpad_GCD(Pointer libHandler, int aMSGIDX, int aTimeOut, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_CEX(Pointer libHandler, boolean VerifyKey, boolean VerifyMagnetic, boolean VerifyICCInsertion, boolean VerifyICCRemoval, boolean VerifyCTLSPresence, int aTimeOut, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_MNU(Pointer libHandler, String sMNUOPT, String sDSPMSG, int aTimeOut, ByteBuffer buffer, IntByReference bufferSize);

        int AbecsPinpad_LoadMedia(Pointer libHandler, String sCaminhoImagem, int aTipoImagem, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_LMF(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
          
        int AbecsPinpad_DSI(Pointer libHandler, String sNomeArquivo);
          
        int AbecsPinpad_DMF(Pointer libHandler, String sNomeArquivo);
    }

    public ACBrAbecsPinpad() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }

        PointerByReference handle = new PointerByReference();
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Inicializar(handle, toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
        setHandle(handle.getValue());
    }

    public ACBrAbecsPinpad(String eArqConfig, String eChaveCrypt) throws Exception {
        PointerByReference handle = new PointerByReference();
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Inicializar(handle, toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
        setHandle(handle.getValue());
    }

    @Override
    protected void dispose() throws Exception {
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Finalizar(getHandle());
        checkResult(ret);
    }

    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Nome(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Versao(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }
    
    public void ConfigImportar(String eArqConfig) throws Exception {

        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_ConfigImportar(getHandle(), eArqConfig);
        checkResult(ret);
    }
    
        public String ConfigExportar() throws Exception {

        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_ConfigExportar(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());
    }

    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_UltimoRetorno(getHandle(), buffer, bufferLen);
    }
    
    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_ConfigLer(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_ConfigGravar(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    @Override
    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_ConfigLerValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    @Override
    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_ConfigGravarValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }

    public void ativar() throws Exception {
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Ativar(getHandle());
        checkResult(ret);
    }

    public void desativar() throws Exception {
        int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_Desativar(getHandle());
        checkResult(ret);
    }

   public void OPN() throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_OPN(getHandle());
       checkResult(ret);
   }
   
   public void CLO(String sMensagem) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_CLO(getHandle(), sMensagem);
       checkResult(ret);
   }
   
   public void CLX(String sMensagemOuNomeImagem) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_CLX(getHandle(), sMensagemOuNomeImagem);
       checkResult(ret);
   }
   
   public String GIX(String PP_DATA) throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_GIX(getHandle(), PP_DATA, buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public String GIN(int GIN_ACQIDX)throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_GIN(getHandle(), GIN_ACQIDX, buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public String PinPadCapabilities() throws Exception{
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_PinPadCapabilities(getHandle(), buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public void DSP(String sMensagem) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_DSP(getHandle(), sMensagem);
       checkResult(ret);
   }
   
   public void DEX(String sMensagem) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_DEX(getHandle(), sMensagem);
       checkResult(ret);
   }
   
   public void GKY() throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_GKY(getHandle());
       checkResult(ret);
   }
   
   public void RMC(String sMensagemRMC) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_RMC(getHandle(), sMensagemRMC);
       checkResult(ret);
   }
   
   public String GCD(int aMSGIDX, int aTimeOut) throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_GCD(getHandle(), aMSGIDX, aTimeOut, buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public String CEX(boolean VerifyKey, boolean VerifyMagnetic, boolean VerifyICCInsertion, boolean VerifyICCRemoval, boolean VerifyCTLSPresence, int aTimeOut) throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_CEX(getHandle(), VerifyKey, VerifyMagnetic, VerifyICCInsertion, VerifyICCRemoval, VerifyCTLSPresence, aTimeOut, buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public String MNU(String sMNUOPT, String sDSPMSG, int aTimeOut) throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_MNU(getHandle(), sMNUOPT, sDSPMSG, aTimeOut, buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public String LoadMedia(String sCaminhoImagem, int aTipoImagem) throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_LoadMedia(getHandle(), sCaminhoImagem, aTipoImagem, buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public String LMF() throws Exception {
       ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
       IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
       
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_LMF(getHandle(), buffer, bufferLen);
       checkResult(ret);
       
       return processResult(buffer, bufferLen);
   }
   
   public void DSI(String sNomeArquivo) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_DSI(getHandle(), sNomeArquivo);
       checkResult(ret);
   }
   
   public void DMF(String sNomeArquivo) throws Exception {
       int ret = ACBrAbecsPinpadLib.INSTANCE.AbecsPinpad_DMF(getHandle(), sNomeArquivo);
       checkResult(ret);
   }
}
